File: plugins/reverb/plate-reverb/PlateReverb.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Plate Reverb (native port) — same Schroeder topology as the JS sibling
// that lived at plate-reverb.js: four parallel feedback combs with a
// one-pole damping LP in each loop, summed and run through two series
// allpasses for diffusion, with a separate pre-delay tap on the input.
//
// The JS version's per-sample inner loop costs ~30 multiply-adds and 6
// array index wraps; running through Rhino's bytecode interpreter that's
// the bulk of CPU on a vocal preview. JVM bytecode collapses the same
// loop to a few hundred nanoseconds per sample.
public final class PlateReverb implements VocalMonitorNativePlugin {

    private float[] preBuf;
    private int preBufLen;
    private int preWrite;

    private final float[][] combs = new float[4][];
    private final int[] combLen = new int[4];
    private final int[] combIdx = new int[4];
    private final float[] combLp = new float[4];
    private final int[] combBase = { 1687, 1601, 2053, 2251 };

    private final float[][] aps = new float[2][];
    private final int[] apLen = new int[2];
    private final int[] apIdx = new int[2];
    private final int[] apBase = { 347, 113 };

    private float srScale = 1f;
    private int sampleRate = 44100;

    private float size = 0.75f, decay = 0.82f, damping = 0.4f, preDelayMs = 0f, mix = 0.32f;

    @Override
    public void init(int sr) {
        this.sampleRate = sr;
        srScale = sr / 44100f;
        for (int c = 0; c < 4; c++) {
            int len = (int) Math.ceil(combBase[c] * 1.6 * srScale);
            combs[c] = new float[len];
            combLen[c] = len;
            combIdx[c] = 0;
            combLp[c] = 0f;
        }
        for (int a = 0; a < 2; a++) {
            int len = (int) Math.ceil(apBase[a] * 1.6 * srScale);
            aps[a] = new float[len];
            apLen[a] = len;
            apIdx[a] = 0;
        }
        preBufLen = (int) Math.ceil(sr * 0.1);
        preBuf = new float[preBufLen];
        preWrite = 0;
    }

    @Override
    public String[] parameterNames() {
        return new String[] { "size", "decay", "damping", "preDelay", "mix" };
    }

    @Override
    public float parameterMin(String name) {
        switch (name) {
            case "size":     return 0.3f;
            case "decay":    return 0f;
            case "damping":  return 0f;
            case "preDelay": return 0f;
            case "mix":      return 0f;
            default:         return 0f;
        }
    }

    @Override
    public float parameterMax(String name) {
        switch (name) {
            case "size":     return 1f;
            case "decay":    return 0.97f;
            case "damping":  return 1f;
            case "preDelay": return 100f;
            case "mix":      return 1f;
            default:         return 1f;
        }
    }

    @Override
    public float parameterDefault(String name) {
        switch (name) {
            case "size":     return 0.75f;
            case "decay":    return 0.82f;
            case "damping":  return 0.4f;
            case "preDelay": return 0f;
            case "mix":      return 0.32f;
            default:         return 0f;
        }
    }

    @Override
    public String parameterLabel(String name) {
        switch (name) {
            case "size":     return "Size";
            case "decay":    return "Decay";
            case "damping":  return "Damping";
            case "preDelay": return "Pre (ms)";
            case "mix":      return "Mix";
            default:         return name;
        }
    }

    @Override
    public void setParameter(String name, float value) {
        switch (name) {
            case "size":     size = value; break;
            case "decay":    decay = value; break;
            case "damping":  damping = value; break;
            case "preDelay": preDelayMs = value; break;
            case "mix":      mix = value; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        final float lpA = 1f - damping * 0.85f;
        final float apG = 0.5f;
        final int preSampClean = Math.max(0, Math.min(preBufLen - 1,
                (int) Math.floor(preDelayMs * sampleRate / 1000f)));

        final int cl0 = Math.max(8, (int) Math.floor(combBase[0] * srScale * (0.6f + 0.4f * size)));
        final int cl1 = Math.max(8, (int) Math.floor(combBase[1] * srScale * (0.6f + 0.4f * size)));
        final int cl2 = Math.max(8, (int) Math.floor(combBase[2] * srScale * (0.6f + 0.4f * size)));
        final int cl3 = Math.max(8, (int) Math.floor(combBase[3] * srScale * (0.6f + 0.4f * size)));
        final int al0 = Math.max(4, (int) Math.floor(apBase[0] * srScale));
        final int al1 = Math.max(4, (int) Math.floor(apBase[1] * srScale));

        final float[] b0 = combs[0], b1 = combs[1], b2 = combs[2], b3 = combs[3];
        final int bl0 = combLen[0], bl1 = combLen[1], bl2 = combLen[2], bl3 = combLen[3];
        int i0 = combIdx[0], i1 = combIdx[1], i2 = combIdx[2], i3 = combIdx[3];
        float l0 = combLp[0], l1 = combLp[1], l2 = combLp[2], l3 = combLp[3];

        final float[] ap0buf = aps[0], ap1buf = aps[1];
        final int ap0Len = apLen[0], ap1Len = apLen[1];
        int ap0i = apIdx[0], ap1i = apIdx[1];
        int pW = preWrite;

        final int n = input.length;
        for (int i = 0; i < n; i++) {
            final float x = input[i];
            preBuf[pW] = x;
            int pR = pW - preSampClean;
            if (pR < 0) pR += preBufLen;
            final float pre = preBuf[pR];
            pW++;
            if (pW >= preBufLen) pW = 0;

            int r0 = i0 - cl0; if (r0 < 0) r0 += bl0;
            float d0 = b0[r0];
            l0 = l0 + lpA * (d0 - l0);
            b0[i0] = pre + l0 * decay;
            i0++; if (i0 >= bl0) i0 = 0;

            int r1 = i1 - cl1; if (r1 < 0) r1 += bl1;
            float d1 = b1[r1];
            l1 = l1 + lpA * (d1 - l1);
            b1[i1] = pre + l1 * decay;
            i1++; if (i1 >= bl1) i1 = 0;

            int r2 = i2 - cl2; if (r2 < 0) r2 += bl2;
            float d2 = b2[r2];
            l2 = l2 + lpA * (d2 - l2);
            b2[i2] = pre + l2 * decay;
            i2++; if (i2 >= bl2) i2 = 0;

            int r3 = i3 - cl3; if (r3 < 0) r3 += bl3;
            float d3 = b3[r3];
            l3 = l3 + lpA * (d3 - l3);
            b3[i3] = pre + l3 * decay;
            i3++; if (i3 >= bl3) i3 = 0;

            float combSum = (l0 + l1 + l2 + l3) * 0.25f;

            int aR0 = ap0i - al0; if (aR0 < 0) aR0 += ap0Len;
            float aD0 = ap0buf[aR0];
            float aIn0 = combSum + aD0 * apG;
            ap0buf[ap0i] = aIn0;
            float ap0Out = aD0 - aIn0 * apG;
            ap0i++; if (ap0i >= ap0Len) ap0i = 0;

            int aR1 = ap1i - al1; if (aR1 < 0) aR1 += ap1Len;
            float aD1 = ap1buf[aR1];
            float aIn1 = ap0Out + aD1 * apG;
            ap1buf[ap1i] = aIn1;
            float wet = aD1 - aIn1 * apG;
            ap1i++; if (ap1i >= ap1Len) ap1i = 0;

            output[i] = x * (1f - mix) + wet * mix;
        }

        combIdx[0] = i0; combIdx[1] = i1; combIdx[2] = i2; combIdx[3] = i3;
        combLp[0] = l0; combLp[1] = l1; combLp[2] = l2; combLp[3] = l3;
        apIdx[0] = ap0i; apIdx[1] = ap1i;
        preWrite = pW;
    }
}

File: plugins/reverb/plate-reverb/plate-reverb.js
// Plate Reverb — Schroeder topology: four parallel feedback comb filters
// (decorrelated prime-ish lengths so their modes don't reinforce into a
// flutter echo) summed and routed through two series allpasses for
// diffusion. Each comb has a one-pole low-pass in its feedback path —
// that's the damping knob: high frequencies decay faster than lows, the
// way a real plate behaves. Classic dense, shimmery 60s/70s plate sound.

function PlateReverbProcessor() {
  // Base comb / allpass lengths in samples. Tuned at 44.1 kHz; the
  // sample-rate factor below scales them so the perceived room size is
  // independent of engine rate. Sized 1.5x to leave room for the "size"
  // knob to push delays longer without re-allocating.
  this.combBase = [1687, 1601, 2053, 2251];
  this.apBase = [347, 113];
  var srScale = sampleRate / 44100;
  this.combs = [];
  this.combLen = [];
  this.combIdx = [0, 0, 0, 0];
  this.combLp = [0, 0, 0, 0];
  for (var c = 0; c < 4; c++) {
    var maxLen = Math.ceil(this.combBase[c] * 1.6 * srScale);
    var b = new Array(maxLen);
    for (var k = 0; k < maxLen; k++) b[k] = 0;
    this.combs.push(b);
    this.combLen.push(maxLen);
  }
  this.allpasses = [];
  this.apLen = [];
  this.apIdx = [0, 0];
  for (var a = 0; a < 2; a++) {
    var maxA = Math.ceil(this.apBase[a] * 1.6 * srScale);
    var ab = new Array(maxA);
    for (var k = 0; k < maxA; k++) ab[k] = 0;
    this.allpasses.push(ab);
    this.apLen.push(maxA);
  }
  this.srScale = srScale;
  this.preBuf = new Array(Math.ceil(sampleRate * 0.1));
  for (var k = 0; k < this.preBuf.length; k++) this.preBuf[k] = 0;
  this.preBufLen = this.preBuf.length;
  this.preWrite = 0;
}

PlateReverbProcessor.parameterDescriptors = [
  { name: 'size',     label: 'Size',      defaultValue: 0.75, minValue: 0.3, maxValue: 1.0 },
  { name: 'decay',    label: 'Decay',     defaultValue: 0.82, minValue: 0,   maxValue: 0.97 },
  { name: 'damping',  label: 'Damping',   defaultValue: 0.4,  minValue: 0,   maxValue: 1 },
  { name: 'preDelay', label: 'Pre (ms)',  defaultValue: 0,    minValue: 0,   maxValue: 100 },
  { name: 'mix',      label: 'Mix',       defaultValue: 0.32, minValue: 0,   maxValue: 1 }
];

PlateReverbProcessor.prototype.process = function (inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var size = parameters.size[0];
  var decay = parameters.decay[0];
  var damping = parameters.damping[0];
  var preMs = parameters.preDelay[0];
  var mix = parameters.mix[0];

  // Damping = how much treble bleeds away per pass. Coefficient form:
  // lpA=0 means full damping (totally lowpassed), lpA=1 means no damping.
  var lpA = 1 - damping * 0.85;
  // Pre-delay tap.
  var preSamp = Math.max(0, Math.min(this.preBufLen - 1, Math.floor(preMs * sampleRate / 1000)));
  // Effective lengths scaled by size knob.
  var cl0 = Math.max(8, Math.floor(this.combBase[0] * this.srScale * (0.6 + 0.4 * size)));
  var cl1 = Math.max(8, Math.floor(this.combBase[1] * this.srScale * (0.6 + 0.4 * size)));
  var cl2 = Math.max(8, Math.floor(this.combBase[2] * this.srScale * (0.6 + 0.4 * size)));
  var cl3 = Math.max(8, Math.floor(this.combBase[3] * this.srScale * (0.6 + 0.4 * size)));
  var al0 = Math.max(4, Math.floor(this.apBase[0] * this.srScale));
  var al1 = Math.max(4, Math.floor(this.apBase[1] * this.srScale));
  var apG = 0.5; // classic Schroeder allpass gain

  for (var i = 0; i < input.length; i++) {
    var x = input[i];
    // Pre-delay line.
    this.preBuf[this.preWrite] = x;
    var preRead = this.preWrite - preSamp;
    if (preRead < 0) preRead += this.preBufLen;
    var pre = this.preBuf[preRead];
    this.preWrite++;
    if (this.preWrite >= this.preBufLen) this.preWrite = 0;

    // Four feedback combs with damping LP inside the loop.
    var b0 = this.combs[0], r0 = this.combIdx[0] - cl0; if (r0 < 0) r0 += this.combLen[0];
    var d0 = b0[r0];
    this.combLp[0] = this.combLp[0] + lpA * (d0 - this.combLp[0]);
    b0[this.combIdx[0]] = pre + this.combLp[0] * decay;
    this.combIdx[0]++; if (this.combIdx[0] >= this.combLen[0]) this.combIdx[0] = 0;

    var b1 = this.combs[1], r1 = this.combIdx[1] - cl1; if (r1 < 0) r1 += this.combLen[1];
    var d1 = b1[r1];
    this.combLp[1] = this.combLp[1] + lpA * (d1 - this.combLp[1]);
    b1[this.combIdx[1]] = pre + this.combLp[1] * decay;
    this.combIdx[1]++; if (this.combIdx[1] >= this.combLen[1]) this.combIdx[1] = 0;

    var b2 = this.combs[2], r2 = this.combIdx[2] - cl2; if (r2 < 0) r2 += this.combLen[2];
    var d2 = b2[r2];
    this.combLp[2] = this.combLp[2] + lpA * (d2 - this.combLp[2]);
    b2[this.combIdx[2]] = pre + this.combLp[2] * decay;
    this.combIdx[2]++; if (this.combIdx[2] >= this.combLen[2]) this.combIdx[2] = 0;

    var b3 = this.combs[3], r3 = this.combIdx[3] - cl3; if (r3 < 0) r3 += this.combLen[3];
    var d3 = b3[r3];
    this.combLp[3] = this.combLp[3] + lpA * (d3 - this.combLp[3]);
    b3[this.combIdx[3]] = pre + this.combLp[3] * decay;
    this.combIdx[3]++; if (this.combIdx[3] >= this.combLen[3]) this.combIdx[3] = 0;

    var combSum = (this.combLp[0] + this.combLp[1] + this.combLp[2] + this.combLp[3]) * 0.25;

    // Two series Schroeder allpasses for diffusion.
    var ap0 = this.allpasses[0], aR0 = this.apIdx[0] - al0; if (aR0 < 0) aR0 += this.apLen[0];
    var aD0 = ap0[aR0];
    var aIn0 = combSum + aD0 * apG;
    ap0[this.apIdx[0]] = aIn0;
    var ap0Out = aD0 - aIn0 * apG;
    this.apIdx[0]++; if (this.apIdx[0] >= this.apLen[0]) this.apIdx[0] = 0;

    var ap1 = this.allpasses[1], aR1 = this.apIdx[1] - al1; if (aR1 < 0) aR1 += this.apLen[1];
    var aD1 = ap1[aR1];
    var aIn1 = ap0Out + aD1 * apG;
    ap1[this.apIdx[1]] = aIn1;
    var wet = aD1 - aIn1 * apG;
    this.apIdx[1]++; if (this.apIdx[1] >= this.apLen[1]) this.apIdx[1] = 0;

    output[i] = x * (1 - mix) + wet * mix;
  }
  return true;
};

registerProcessor('plate-reverb', PlateReverbProcessor);
